keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -0,0 +1,198 @@
+  * Set up a CRD, persist some data, changed the version, and ensure the previously persisted data is readable
+  * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
+  * Ensure `spec.served` is respected
+
+### e2e tests for GA
+
+* Custom Resources should be readable and writable at all available versions (test for round trip-ability)
+* A Custom Resource watch should terminate if its CustomResourceDefinition is deleted or updated
+
+### Scale Targets for GA
+
+* TODO quantify: Read/write latency of CRDs within X% of native Kubernetes types
+* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
+* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
+* Coordinate with sig-scalability
+
+## Graduation Criteria
+
+To mark these as complete, all of the above features need to be implemented. An
+umbrella issue is tracking all of these changes. Also there need to be
+sufficient tests for any of these new features and all existing features and
+documentation should be completed for all features.
+
+See [umbrella issue](https://github.com/kubernetes/kubernetes/issues/58682) for status.
+
+## Post-GA tasks
+
+See the [umbrella issue](https://github.com/kubernetes/kubernetes/issues/58682) for details on Post-GA tasks. The tasks at the time this KEP was written are:
+
+* Human readable status from conditions for a CRD using additionalPrinterColumns (https://github.com/kubernetes/kubernetes/issues/67268)
+* Consider changing the schema language in CRDs (https://github.com/kubernetes/kubernetes/issues/67840)
+* Should support graceful deletion for storage (https://github.com/kubernetes/kubernetes/issues/68508)
+* Enable arbitrary CRD field selectors by supporting a whitelist of fields in CRD spec (https://github.com/kubernetes/kubernetes/issues/53459)
+* Support graceful deletion for custom resources (https://github.com/kubernetes/kubernetes/issues/56567)
+* CRD validation webhooks (https://github.com/kubernetes/community/pull/1418)
+* Allow OpenAPI references in the CRD valiation schema (https://github.com/kubernetes/kubernetes/issues/54579, https://github.com/kubernetes/kubernetes/issues/62872)
+* Generate json-schema for use in the CRDs from the go types (https://github.com/kubernetes/kubernetes/issues/59154, https://github.com/kubernetes/sample-controller/issues/2, https://github.com/kubernetes/code-generator/issues/28)
+* Support for namespace-local CRD (https://github.com/kubernetes/kubernetes/issues/65551)
+* Support proto encoding for custom resources (https://github.com/kubernetes/kubernetes/issues/63677)
+* labelSelectorPath should be allowed not be under .status (https://github.com/kubernetes/kubernetes/issues/66688)
+* Support arbitrary non-standard subresources for CRDs (https://github.com/kubernetes/kubernetes/issues/72637)
+* OpenAPI v3 is published for all resources, including custom resources
```

Comment:
We have found some client issues dealing with unexpected openapi conditions during our alpha stage.  I'm concerned about promoting to GA without seeing this be functional because we're already talking about stripping back the set of supported features and I don't want to make GA objects invalid.

Reply:
This is talking about publishing openapi v3, which we don't do for any resources today, including in-tree ones. The [openapi subset KEP](https://github.com/kubernetes/enhancements/pull/1002) is intended to ensure we limit ourselves to what can be published for openapi v3.